### PR TITLE
TRUNK-6070 Add Patient Program to Patient Identifier

### PR DESCRIPTION
**Pull request by slubwama**
Ticket
https://issues.openmrs.org/browse/TRUNK-6070
Add a patient program to the patient identifier to create a connection between a patient identifier and a patient program
Create a method to allow getting a patient identifier by a patient program.

SAME AS https://github.com/openmrs/openmrs-core/pull/4042

### CODE REVIEW

**Comment by dkayiwa** [comment]
Did you see the above maven build failure? The liquibase changes should be in `liquibase-update-to-latest-2.6.x.xml`

File: api/src/main/resources/org/openmrs/liquibase/snapshots/core-data/liquibase-core-data-2.6.x.xml [hunk 1]
```diff
@@ -0,0 +1,16 @@
+<?xml version="1.0" encoding="UTF-8"?>
+
+<!--
+
+    This Source Code Form is subject to the terms of the Mozilla Public License,
+    v. 2.0. If a copy of the MPL was not distributed with this file, You can
+    obtain one at http://mozilla.org/MPL/2.0/. OpenMRS is also distributed under
+    the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
+
```
Comment (dkayiwa) [inline api/src/main/resources/org/openmrs/liquibase/snapshots/core-data/liquibase-core-data-2.6.x.xml:9]: @slubwama why did you add this file?

File: api/src/main/resources/org/openmrs/liquibase/snapshots/schema-only/liquibase-schema-only-2.6.x.xml [hunk 2]
```diff
@@ -11,7 +10,15 @@
     graphic logo is a trademark of OpenMRS Inc.
 
 -->
-<databaseChangeLog xmlns="http://www.liquibase.org/xml/ns/dbchangelog" xmlns:ext="http://www.liquibase.org/xml/ns/dbchangelog-ext" xmlns:pro="http://www.liquibase.org/xml/ns/pro" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog-ext http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-ext.xsd http://www.liquibase.org/xml/ns/pro http://www.liquibase.org/xml/ns/pro/liquibase-pro-3.10.xsd http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.10.xsd">
+<databaseChangeLog
+	xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
+	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	xmlns:ext="http://www.liquibase.org/xml/ns/dbchangelog-ext"
+	xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog-ext
```
Comment (dkayiwa) [inline api/src/main/resources/org/openmrs/liquibase/snapshots/schema-only/liquibase-schema-only-2.6.x.xml:17]: Why did you make changes to this file?

Comment (slubwama) [reply api/src/main/resources/org/openmrs/liquibase/snapshots/schema-only/liquibase-schema-only-2.6.x.xml:17]: @dkayiwa when I checked with the failing test about liquidbase it was pointing to this specific file. After making the change the test passed. How then do you add version series

Comment (dkayiwa) [reply api/src/main/resources/org/openmrs/liquibase/snapshots/schema-only/liquibase-schema-only-2.6.x.xml:17]: Like i said above, your changes should be in `liquibase-update-to-latest-2.6.x.xml`

**Comment by dkayiwa** [comment]
@slubwama from this failure log `DatabaseUpdaterDatabaseIT.should:55 expected: <891> but was: <892>` you also need to increment `DatabaseUpdaterDatabaseIT` by the number of changesets that you have added.

**Comment by slubwama** [comment]
@dkayiwa done with the changes

File: api/src/main/java/org/openmrs/PatientIdentifier.java [hunk 3]
```diff
@@ -328,4 +331,21 @@ public int compare(PatientIdentifier pi1, PatientIdentifier pi2) {
 			return retValue;
 		}
 	}
+
+
+	/**
+	 * @since 2.3.6
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/PatientIdentifier.java:337]: It would be great to add a description to this method.

File: api/src/main/java/org/openmrs/PatientIdentifier.java [hunk 4]
```diff
@@ -328,4 +331,21 @@ public int compare(PatientIdentifier pi1, PatientIdentifier pi2) {
...
+	 * @return patientProgram
+	 */
+	public PatientProgram getPatientProgram() {
+		return patientProgram;
+	}
+
+	/**
+	 * @since 2.3.6
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/PatientIdentifier.java:345]: It would be great to add a description to this method.

File: api/src/main/java/org/openmrs/api/PatientService.java [hunk 5]
```diff
@@ -844,4 +845,14 @@ public void saveCauseOfDeathObs(Patient patient, Date dateDied, Concept causeOfD
 	 * @throws PatientIdentifierTypeLockedException
 	 */
 	public void checkIfPatientIdentifierTypesAreLocked() throws PatientIdentifierTypeLockedException;
-}
\ No newline at end of file
+
+	/**
+	 * Get all patientIdentifiers that are associated to the patient program
+	 * returned
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/api/PatientService.java:851]: What does the above `returned` mean?

File: api/src/main/java/org/openmrs/api/db/PatientDAO.java [hunk 6]
```diff
@@ -296,5 +297,11 @@
 	 * @return the saved allergy
 	 */
 	public Allergy saveAllergy(Allergy allergy);
+
+	/**
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/api/db/PatientDAO.java:301]: It would be great to add a description to this method.

Comment (slubwama) [reply api/src/main/java/org/openmrs/api/db/PatientDAO.java:301]: @dkayiwa done

File: api/src/main/java/org/openmrs/api/PatientService.java (hunk 5 above)
Comment (slubwama) [reply api/src/main/java/org/openmrs/api/PatientService.java]: @dkayiwa  removed

File: api/src/main/java/org/openmrs/PatientIdentifier.java (hunk 4 above)
Comment (slubwama) [reply api/src/main/java/org/openmrs/PatientIdentifier.java:347]: @dkayiwa done

File: api/src/main/java/org/openmrs/PatientIdentifier.java (hunk 3 above)
Comment (slubwama) [reply api/src/main/java/org/openmrs/PatientIdentifier.java:338]: @dkayiwa done

File: api/src/main/java/org/openmrs/PatientIdentifier.java [hunk 7]
```diff
@@ -328,4 +331,23 @@ public int compare(PatientIdentifier pi1, PatientIdentifier pi2) {
 			return retValue;
 		}
 	}
+
+
+	/**
+	 * This method returns a patient program associated to the identifier in context
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/PatientIdentifier.java:337]: We tend to avoid what is already obvious. For instance, everyone knows that this is a method. :)

File: api/src/main/java/org/openmrs/PatientIdentifier.java [hunk 8]
```diff
@@ -328,4 +331,23 @@ public int compare(PatientIdentifier pi1, PatientIdentifier pi2) {
...
+	public PatientProgram getPatientProgram() {
+		return patientProgram;
+	}
+
+	/**
+	 * This method sets the patient program on a patient Identifier
+	 * @since 2.3.6
+	 * @param patientProgram The patientProgram to set.
+	 */
+	public void setPatientProgram(PatientProgram patientProgram) {
```
Comment (tendomart) [inline api/src/main/java/org/openmrs/PatientIdentifier.java:350]: This would also be good to be like  @param patientProgram The patient program to set.

File: api/src/main/java/org/openmrs/api/PatientService.java [hunk 9]
```diff
@@ -844,4 +845,13 @@ public void saveCauseOfDeathObs(Patient patient, Date dateDied, Concept causeOfD
 	 * @throws PatientIdentifierTypeLockedException
 	 */
 	public void checkIfPatientIdentifierTypesAreLocked() throws PatientIdentifierTypeLockedException;
-}
\ No newline at end of file
+
+	/**
+	 * Get all patientIdentifiers that are associated to the patient program
+	 * @param patientProgram the patientProgram to be used to fetch the associated identifiers
```
Comment (tendomart) [inline api/src/main/java/org/openmrs/api/PatientService.java:851]: * @param patientProgram the patient program to be used to fetch the associated identifiers
